compiler/rustc_codegen_llvm/src/base.rs

```diff
@@ -109,6 +109,14 @@ pub(crate) fn compile_codegen_unit(
                 attributes::apply_to_llfn(entry, llvm::AttributePlace::Function, &attrs);
             }
 
+            // Define Objective-C module info for 32-bit x86 macOS. Note, this generates a global
+            // that gets added to the `llvm.compiler.used` variable, created later.
+            let uses_objc =
+                !cx.objc_classrefs.borrow().is_empty() || !cx.objc_selrefs.borrow().is_empty();
```

Comment:
An important note that might be useful to capture in a code comment: The module info / module flags do not have any effect on non-Objective-C code / statics. It is only a signal to the linker that it needs to do more work on object files that contain this.

That is, it would be valid to always emit it, but we choose not to, for slightly better link performance.